File: Sort/Bubble/bubbleSort.py
class BubbleSort(object):
    def Sort(self, nums):
        swapped = True
        while (swapped):
            swapped = False
            for i in range(1, len(nums) - 1):
                if nums[i] < nums[i - 1]:
                    tmp = nums[i - 1]
                    nums[i - 1] = nums[i]
                    nums[i] = tmp
                    swapped = True
        return nums

    def Sort2(self, nums):
        for i in range(len(nums)):
            for j in range(1 + i, len(nums) - 1):
                if nums[j] < nums[j - 1]:
                    tmp = nums[j - 1]
                    nums[j - 1] = nums[j]
                    nums[j] = tmp

        return nums


    def Sort3(self, nums):
        swapped = True
        while swapped:
            swapped = False
            for i in range(len(nums) - 2, -1, -1):
                if nums[i + 1] > nums[i]:
                    tmp = nums[i + 1]
                    nums[i + 1] = nums[i]
                    nums[i] = tmp
                    swapped = True

        return nums

    def Sort4(self, nums):
        for i in range(len(nums) - 1):
            for j in range(len(nums) - 2, -1 + i, -1):
                if nums[j + 1] > nums[j]:
                    tmp = nums[j + 1]
                    nums[j + 1] = nums[j]
                    nums[j] = tmp
        
        return nums

bs = BubbleSort()
nums = [7, 6, 9, 8, 7, 6, 3, 7, 5, 5, 7, 8, 9, 7, 8]
print(bs.Sort(nums))
print(bs.Sort2(nums))
print(bs.Sort3(nums))
print(bs.Sort4(nums))

File: Sort/Bubble/bubbleSort.java
class BubbleSort {
  BubbleSort() {

  }

  // asc
  public int[] Sort(int[] nums) {
    Boolean swapped = true;
    while (swapped) {
      swapped = false;
      for (int i = 1; i < nums.length; i++) {
        if (nums[i - 1] > nums[i]) {
          int tmp = nums[i - 1];
          nums[i - 1] = nums[i];
          nums[i] = tmp;
          swapped = true;
        }
      }
    }
    return nums;
  }

  public int[] Sort2(int[] nums) {
    int size = nums.length;
    for (int i = 0; i < size - 1; i++) {
      for (int j = 1; j < size - i; j++) {
        if (nums[j - 1] > nums[j]) {
          int tmp = nums[j - 1];
          nums[j - 1] = nums[j];
          nums[j] = tmp;
        }
      }
    }
    return nums;
  }

  // desc
  public int[] Sort3(int[] nums) {
    Boolean swapped = true;
    while (swapped) {
      swapped = false;
      for (int i = nums.length - 2; i >= 0; i--) {
        if (nums[i] < nums[i + 1]) {
          int tmp = nums[i + 1];
          nums[i + 1] = nums[i];
          nums[i] = tmp;
          swapped = true;
        }
      }
    }
    return nums;
  }

  public int[] Sort4(int[] nums) {
    int size = nums.length;
    for (int i = 0; i < size; i++) {
      for (int j = size - 2; j >= i; j--) {
        if (nums[j] < nums[j + 1]) {
          int tmp = nums[j];
          nums[j] = nums[j + 1];
          nums[j + 1] = tmp;
        }
      }
    }
    return nums;
  }

  public static void main(String[] args) {
    int[] nums = {7, 6, 9, 8, 7, 6, 3, 7, 5, 5, 7, 8, 9, 7, 8};
    BubbleSort bs = new BubbleSort();
    int[] sorted = bs.Sort(nums);
    for (int i : sorted) {
      System.out.print(i);
    }
    System.out.println("");
    System.out.println("=========================");
    int[] sorted2 = bs.Sort2(nums);
    for (int i : sorted2) {
      System.out.print(i);
    }
    System.out.println("");
    System.out.println("=========================");
    int[] sorted3 = bs.Sort3(nums);
    for (int i : sorted3) {
      System.out.print(i);
    }
    System.out.println("");
    System.out.println("=========================");
    int[] sorted4 = bs.Sort4(nums);
    for (int i : sorted4) {
      System.out.print(i);
    }
    System.out.println("");
  }
}